Add Vulkan job dispatch and flush.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44017 Add Vulkan op Add.
* #46481 Add Vulkan tensor copy.
* #44016 Add Vulkan Tensor factory.
* #44015 Add Vulkan Tensor.
* #46503 Provide CMake option to enable Vulkan API.
* **#46008 Add Vulkan job dispatch and flush.**

Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

Commit:
Add Vulkan job dispatch and flush.

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Command.cpp:
```diff
@@ -242,17 +242,21 @@ void Command::Buffer::copy(
       "This command buffer is in an invalid state! "
       "Potential reason: This command buffer is moved from.");
 
+  const auto div_round_up = [](const uint32_t n, const uint32_t d) {
```
could it be static to create it only once?

AshkanAliabadi replied:
Great point, will do.

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Context.cpp:
```diff
@@ -78,25 +78,42 @@ VkQueue acquire_queue(
+  try {
+    flush();
+  }
+  catch (...) {
```
Is it by design that it will catch everything silently even for debug?

Thinking that at least for debug we might want to see the errors or will they show up at some other listener?

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Context.h:
```diff
@@ -26,10 +26,10 @@ class Context final {
   explicit Context(const Adapter& adapter);
   Context(const Context&) = delete;
-  Context(Context&&) = default;
+  Context(Context&&) = delete;
```
Why do you want to make it non-movable?

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan job dispatch and flush."


Differential Revision: [D24291507](https://our.internmc.facebook.com/intern/diff/D24291507)

[ghstack-poisoned]
